### PR TITLE
[Merged by Bors] - feat: show that `BitVec` is a commutative ring

### PR DESCRIPTION
**Pull request by alexkeizer**
---

- [x] depends on: #8700 [That PR breaks this one]

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Bitvec/Lemmas.lean
Comment (eric-wieser) [inline Mathlib/Data/Bitvec/Lemmas.lean:165]: I think it would be better to use `Function.Injective.commRing`.

File: Mathlib/Data/Bitvec/Lemmas.lean
Comment (eric-wieser) [reply Mathlib/Data/Bitvec/Lemmas.lean:140]: (if the user wanted to work with `Fin` instead, then they could apply `toFin_injective`, and push `toFin` to the inside in the same way).

Comment (alexkeizer) [reply Mathlib/Data/Bitvec/Lemmas.lean:140]: My reason for the direction is that `ofFin x &&& ofFin y` unfolds to `(ofFin x).toFin &&& (ofFin y).toFin`, where `(ofFin x).toFin = x`. So, the `Fin` expression is (in my mind) simpler / more fundamental. Turning that aroung seems to me not something we'd want `simp` to do by default.

I'll concede that I often forgot how useful `toFin_injective` can be, but it cannot always be used. For example, for inequalities or in subexpressions. Then again, in my current use-case it can, so I'm more than happy to leave these theorems out if you insist they are not desirable

Comment (eric-wieser) [reply Mathlib/Data/Bitvec/Lemmas.lean:140]: I'd prefer that we keep the *reverse* of these theorems; they match things like `MulOpposite.op_add` etc.

File: Mathlib/Data/Bitvec/Lemmas.lean
Comment (alexkeizer) [reply Mathlib/Data/Bitvec/Lemmas.lean:165]: That seems cleaner, yes. However, `Function.Injective.commRing` requires instances of `[SMul ℕ β] [SMul ℤ β] [Pow β ℕ]`, which `CommRing` by itself does not need.
Is this by design, it seems `Function.Injective.commRing` shouldn't need them if it's just defining an instance of `CommRing`?

Comment (eric-wieser) [reply Mathlib/Data/Bitvec/Lemmas.lean]: > which CommRing by itself does not need.

Actually it does need them, but it generates them automatically if you forget them.
The point here is that probably you don't want the auto-generated ones; the manual implementations in terms of `Fin` will likely have better properties.

File: Mathlib/Data/Bitvec/Lemmas.lean
Comment (eric-wieser) [inline Mathlib/Data/Bitvec/Lemmas.lean:35]: nit: I'd put these before the `toNat` ones, as they're in some sense more fundamental. You could also then replace the proof of `Function.Injective (@Std.BitVec.toNat n)` with `toFin_injective.comp Fin.val_injective` or similar if you wanted, not that it really matters.

File: Mathlib/Data/Bitvec/Lemmas.lean
Comment (eric-wieser) [inline Mathlib/Data/Bitvec/Lemmas.lean:146]: These names aren't quite right:
```suggestion
@[simp] lemma ofFin_neg (x : Fin (2^w)) : ofFin (-x) = -(ofFin x) := by
  rw [neg_eq_zero_sub]; rfl

@[simp] lemma ofFin_and (x y : Fin (2^w)) : ofFin (x &&& y) = ofFin x &&& ofFin y := rfl
@[simp] lemma ofFin_or  (x y : Fin (2^w)) : ofFin (x ||| y) = ofFin x ||| ofFin y := rfl
@[simp] lemma ofFin_xor (x y : Fin (2^w)) : ofFin (x ^^^ y) = ofFin x ^^^ ofFin y := rfl
@[simp] lemma ofFin_add (x y : Fin (2^w)) : ofFin (x + y)   = ofFin x + ofFin y   := rfl
@[simp] lemma ofFin_sub (x y : Fin (2^w)) : ofFin (x - y)   = ofFin x - ofFin y   := rfl
@[simp] lemma ofFin_mul (x y : Fin (2^w)) : ofFin (x * y)   = ofFin x * ofFin y   := rfl
```

File: Mathlib/Data/Bitvec/Lemmas.lean
Comment (eric-wieser) [inline Mathlib/Data/Bitvec/Lemmas.lean:149]:
```suggestion
## Distributivity of `Std.BitVec.toFin`
```
so that this is linked in the docs.

File: Mathlib/Data/Bitvec/Lemmas.lean
Comment (eric-wieser) [inline Mathlib/Data/Bitvec/Lemmas.lean:163]: I'd recommend 
```suggestion
  smul x y :=.ofFin (x • y.toFin)
```
and similarly below.

File: Mathlib/Data/Bitvec/Lemmas.lean
Comment (eric-wieser) [inline Mathlib/Data/Bitvec/Lemmas.lean:159]:
```suggestion
@[simp] lemma ofFin_zero : ofFin 0 = 0 := rfl
@[simp] lemma ofFin_one : ofFin 1 = 1 := rfl
```
though these probably belong in the section above.

File: Mathlib/Data/Bitvec/Lemmas.lean
Comment (eric-wieser) [inline Mathlib/Data/Bitvec/Lemmas.lean:187]: rather than `intros; try rfl`, can you use the names of the lemmas that you already created above? It's a nice check that you wrote all the ones that matter.

File: Mathlib/Data/Bitvec/Lemmas.lean
Comment (eric-wieser) [inline Mathlib/Data/Bitvec/Lemmas.lean:176]: I guess these two shouldn't be `simp`

File: Mathlib/Data/Bitvec/Lemmas.lean
Comment (eric-wieser) [inline Mathlib/Data/Bitvec/Lemmas.lean:170]: A nicer way to write this is
```suggestion
@[simp] lemma toFin_zero    : toFin (0 : BitVec n) = 0 := rfl
```

File: Mathlib/Data/Bitvec/Lemmas.lean
Comment (eric-wieser) [inline Mathlib/Data/Bitvec/Lemmas.lean:171]: This looks like a bad choice of simp lemma by std, so I'd be inclined to leave a TODO comment (after removing `simp`)

File: Mathlib/Data/Bitvec/Lemmas.lean
Comment (eric-wieser) [inline Mathlib/Data/Bitvec/Lemmas.lean:179]: Can you add the `OfNat` lemma too? It will probably need to be after the ring instance

Comment (eric-wieser) [reply Mathlib/Data/Bitvec/Lemmas.lean]: I added these myself

**Comment by eric-wieser** [comment]
maintainer merge

I pushed some cleanups, so someone should give this a final look over

File: Mathlib/Data/Bitvec/Lemmas.lean
Comment (alexkeizer) [reply Mathlib/Data/Bitvec/Lemmas.lean]: Thanks!

**Comment by alexkeizer** [comment]
This PR will probably need to wait, I just noticed the std bump in #8700 changes a lot of the BitVector definitions and breaks some of the trivial proofs `rfl` here. Since #8700 seems to be higher priority, I propose we wait until that has been merged, then fix this PR.

File: Mathlib/Data/BitVec/Defs.lean
Comment (eric-wieser) [inline Mathlib/Data/BitVec/Defs.lean:149]: Using `.ofInt` probably is the right thing to do, since ultimately `ofInt` is likely to be more efficient.

Comment (alexkeizer) [reply Mathlib/Data/BitVec/Defs.lean]: Fair point, I've redefined `IntCast` and proven the necessary equivalence with `Fin`s `IntCast` instance

**Comment by alexkeizer** [comment]
#8700 has been merged, and this PR fixed accordingly, so it is once more ready for review!

**Comment by tobiasgrosser** [comment]
I guess this may have fallen from the maintainer queue due to @alexkeizer cancelling the merge. @semorrison, any chance you could have a brief look?

File: Mathlib/Data/BitVec/Lemmas.lean
Comment (eric-wieser) [inline Mathlib/Data/BitVec/Lemmas.lean:154]: Do you have any opinions on leanprover/std4#415? That will almost certainly break this proof, so it probably makes sense for you to review it (and maybe wait for a merge) before merging this.

Comment (alexkeizer) [reply Mathlib/Data/BitVec/Lemmas.lean:154]: It's not immediately obvious to me that your version indeed is faster, but otherwise that PR LGTM.

I would prefer we merge this one sooner rather than later, though, since the various `toFin_` and `ofFin_` lemmas are also useful for further developments. Would it be a problem to just merge this now, and fix the proof in the std bump PR once leanprover/std4#415 gets merged? 
Alternatively, I can replace this proof with a `proof_wanted`, comment out whatever depends on it (i.e., the `CommRing` instance), and factor those out to a new PR, so that atleast those lemmas are available.

Comment (tobiasgrosser) [reply Mathlib/Data/BitVec/Lemmas.lean:154]: https://github.com/leanprover/std4/pull/415 is now merged into master. I rebased this PR on master, such that we now can rework this proof.

File: Mathlib/Data/BitVec/Lemmas.lean
Comment (eric-wieser) [inline Mathlib/Data/BitVec/Lemmas.lean:274]: Perhaps it makes sense to reduce this PR to the `CommSemiring` case to get it merged, and then come back and prove the necessary things for `CommRing` about `ofInt` later?

Comment (alexkeizer) [reply Mathlib/Data/BitVec/Lemmas.lean]: Agreed, I've parked the attempt at a proof I did earlier in the `bitvec-ring2` branch, and reduced this PR as suggested

File: Mathlib/Data/BitVec/Defs.lean
Comment (eric-wieser) [inline Mathlib/Data/BitVec/Defs.lean:146]: I think this instance is already defined by Std? Maybe I'm mistaken

File: Mathlib/Data/BitVec/Lemmas.lean
Comment (eric-wieser) [inline Mathlib/Data/BitVec/Lemmas.lean:214]:
```suggestion
-- TODO: generalize to `CommRing` after `ofFin_intCast` is proven
```

File: Mathlib/Data/BitVec/Lemmas.lean
Comment (eric-wieser) [inline Mathlib/Data/BitVec/Lemmas.lean:208]: Perhaps worth a comment that either of these follows trivially from the other (though that doesn't help with deciding which one to prove)

File: Mathlib/Data/Nat/Order/Lemmas.lean
Comment (eric-wieser) [inline Mathlib/Data/Nat/Order/Lemmas.lean:281]: This lemma looks quite strange to me; I think it would be best left to a later PR

File: Mathlib/Data/BitVec/Lemmas.lean
Comment (tobiasgrosser) [reply Mathlib/Data/BitVec/Lemmas.lean:210]: Done.

File: Mathlib/Data/Nat/Order/Lemmas.lean
Comment (tobiasgrosser) [reply Mathlib/Data/Nat/Order/Lemmas.lean]: Dropped.

File: Mathlib/Data/BitVec/Lemmas.lean
Comment (eric-wieser) [inline Mathlib/Data/BitVec/Lemmas.lean:207]:
```suggestion
-- Either of these follows trivially from the other. Which one to
-- prove is not yet clear.
```

File: Mathlib/Data/BitVec/Defs.lean
Comment (tobiasgrosser) [reply Mathlib/Data/BitVec/Defs.lean:146]: These are all the instances that are defined in https://github.com/leanprover/std4/commit/fa17cba0459830906e99f0a81fb0edc106220380:

```bash
grep -R instance Std/Data/BitVec
Std/Data/BitVec/Basic.lean:instance : Inhabited (BitVec n) where default := .zero n
Std/Data/BitVec/Basic.lean:instance : OfNat (BitVec n) i where ofNat := .ofNat n i
Std/Data/BitVec/Basic.lean:instance : Repr (BitVec n) where reprPrec a _ := "0x" ++ (a.toHex : Format) ++ "#" ++ repr n
Std/Data/BitVec/Basic.lean:instance : ToString (BitVec n) where toString a := toString (repr a)
Std/Data/BitVec/Basic.lean:instance : Add (BitVec n) := ⟨BitVec.add⟩
Std/Data/BitVec/Basic.lean:instance : Sub (BitVec n) := ⟨BitVec.sub⟩
Std/Data/BitVec/Basic.lean:instance : Neg (BitVec n) := ⟨.neg⟩
Std/Data/BitVec/Basic.lean:instance : Mul (BitVec n) := ⟨.mul⟩
Std/Data/BitVec/Basic.lean:instance : Div (BitVec n) := ⟨.udiv⟩
Std/Data/BitVec/Basic.lean:instance : Mod (BitVec n) := ⟨.umod⟩
Std/Data/BitVec/Basic.lean:instance : LT (BitVec n) where lt x y := x.toFin < y.toFin
Std/Data/BitVec/Basic.lean:instance : LE (BitVec n) where le x y := x.toFin ≤ y.toFin
Std/Data/BitVec/Basic.lean:instance : AndOp (BitVec w) := ⟨.and⟩
Std/Data/BitVec/Basic.lean:instance : OrOp (BitVec w) := ⟨.or⟩
Std/Data/BitVec/Basic.lean:instance : Xor (BitVec w) := ⟨.xor⟩
Std/Data/BitVec/Basic.lean:instance : Complement (BitVec w) := ⟨.not⟩
Std/Data/BitVec/Basic.lean:instance : HShiftLeft (BitVec w) Nat (BitVec w) := ⟨.shiftLeft⟩
Std/Data/BitVec/Basic.lean:instance : HShiftRight (BitVec w) Nat (BitVec w) := ⟨.ushiftRight⟩
Std/Data/BitVec/Basic.lean:instance {n} : HShiftLeft  (BitVec m) (BitVec n) (BitVec m) := ⟨fun x y => x <<< y.toNat⟩
Std/Data/BitVec/Basic.lean:instance {n} : HShiftRight (BitVec m) (BitVec n) (BitVec m) := ⟨fun x y => x >>> y.toNat⟩
Std/Data/BitVec/Basic.lean:instance : HAppend (BitVec w) (BitVec v) (BitVec (w + v)) := ⟨.append⟩
Std/Data/BitVec/Basic.lean:instance : Subsingleton (BitVec 0) where
```

I do not see NatCast.

Comment (alexkeizer) [reply Mathlib/Data/BitVec/Defs.lean:146]: I've confirmed with `#synth` that the instances at the least don't exist in the files imported at that point, which includes `Std.Data.BitVec`, so I'm pretty sure they don't exist in Std at all

File: Mathlib/Data/BitVec/Lemmas.lean
Comment (eric-wieser) [inline Mathlib/Data/BitVec/Lemmas.lean:8]: I don't see any uses of `linarith` in this file.

**Review by eric-wieser — APPROVED** [review]
bors d+

Thanks!

**Comment by alexkeizer** [comment]
Thanks for the quick final review @eric-wieser!

File: Mathlib/Data/BitVec/Lemmas.lean
Comment (eric-wieser) [reply Mathlib/Data/BitVec/Lemmas.lean]: Can you open a new WIP PR from that branch so that it doesn't get forgotten?

**Pull request closed without merging**
